jax/_src/nn/functions.py

```diff
@@ -765,3 +767,174 @@ def hard_silu(x: ArrayLike) -> Array:
+    is_causal: If true, assumes upper left causal attention masking.
+    implementation: A string to control which implementation to use. Supported
+                    strings are `xla` (default), `cudnn` (cuDNN flash
+                    attention).
+
+  Returns:
+    An array with the same shape of :code:`query`. If flash attention is not
+    used, the `probs` array will also be returned.
+  """
+  def _assert_has_shape(t: ArrayLike, shape: Sequence[int]) -> None:
+    assert t.ndim == len(shape)
+    value_str1 = f't.shape={t.shape}'
+    value_str2 = f'shape={shape}'
+    for i in range(t.ndim):
+      if shape[i] != -1:
+        assert (
+            t.shape[i] == shape[i]
+        ), f'Shapes are not equal {value_str1} vs {value_str2}'
+    
+  B, S, N, H = key.shape
+  _assert_has_shape(value, [B, S, N, H])
+  _assert_has_shape(query, [B, -1, N, H])
+  T = query.shape[1]
+  scale_val = (1.0 / np.sqrt(H) if scale is None else scale)
+  assert query.dtype == key.dtype == value.dtype
+  if mask is not None:
+    assert mask.dtype == query.dtype or mask.dtype == jnp.bool_
+
+  if implementation in ('xla', None):
+    encoded, probs = _dot_product_attention_xla(
+        query, key, value, bias, mask, is_causal=is_causal, scale=scale_val
+    )
+  elif implementation == 'cudnn':
+    mask_type = MaskType.CAUSAL if is_causal else MaskType.NO_MASK
+    # Convert bool mask to float mask for addition
+    if mask is not None:
+      if mask.dtype == jnp.bool_:
+        large_negative_number = _get_large_negative(query.dtype)
+        mask = jnp.where(mask, jnp.zeros((), query.dtype),
+                         large_negative_number)
+
+    # Prepare the bias for cudnn flash attention:
+    # (1) We should never use the mask argument of cudnn, because it is
+    #     multiplicative and thus the masked values (i.e. the zeros) will
+    #     still take part in the following softmax. So, we need to use the bias
+    #     argument for the mask to ensure the masked values are very small.
+    # (2) Cudnn does the scale before the bias addition, i.e. scale*QK+bias.
+    #     However, we'd like the bias addition before the scale, i.e.
+    #     scale*(QK+bias). So, we scale the bias beforehand.
+    # TODO(kaixih@nvidia): The logic of (1) should be moved to the internal of
+    # cudnn_dot_product_attention. For (2), it seems we want to support both
+    # cases, e.g., flax/cudnn attention is the post-scale bias, but praxis
+    # attention is the post-bias scale.
+    if bias is None:
+      bias = mask
+    else:
+      scale = jnp.array(scale_val, dtype=bias.dtype)
+      bias = jnp.multiply(bias, scale)
+      if mask is not None:
+        bias = bias + mask
+
+    encoded = cudnn_dot_product_attention(
+        query, key, value, bias, mask=None, scale=scale_val, mask_type=mask_type
+    )
+    probs = None
+  else:
+    raise ValueError(f"Unsupported implementation option: {implementation}")
+
+  return encoded, probs
```

Comment:
This does not match the type signature of the function.

Reply:
Done.